server/src/main/java/io/crate/expression/scalar/regex/RegexMatcher.java

```diff
@@ -22,70 +22,37 @@
 package io.crate.expression.scalar.regex;
 
 import javax.annotation.Nullable;
-import java.util.ArrayList;
 import java.util.List;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
-public class RegexMatcher {
+public abstract class RegexMatcher {
 
-    private final boolean globalFlag;
     private final Pattern pattern;
     private Matcher matcher;
 
-    public RegexMatcher(String regex, int flags, boolean globalFlag) {
-        pattern = Pattern.compile(regex, flags);
-        this.globalFlag = globalFlag;
-    }
-
     public RegexMatcher(String regex, @Nullable String flags) {
-        this(regex, parseFlags(flags), isGlobal(flags));
-    }
-
-    public RegexMatcher(String regex) {
-        this(regex, 0, false);
-    }
-
-    public boolean match(String term) {
-        initMatcher(term);
-        return matcher.find();
-    }
-
-    @Nullable
-    public List<String> groups() {
-        int groupCount = matcher.groupCount();
-        if (groupCount == 0) {
-            try {
-                return List.of(matcher.group());
-            } catch (IllegalStateException e) {
-                return null;
-            }
-        }
-        ArrayList<String> groups = new ArrayList<>(groupCount);
-        for (int i = 0; i < groupCount; i++) {
-            groups.add(matcher.group(i + 1));
-        }
-        return groups;
+        pattern = Pattern.compile(regex, parseFlags(flags));
     }
 
-    public String replace(String term, String replacement) {
-        initMatcher(term);
-        if (globalFlag) {
-            return matcher.replaceAll(replacement);
-        } else {
-            return matcher.replaceFirst(replacement);
-        }
-    }
-
-    private void initMatcher(String term) {
+    protected Matcher initMatcher(String term) {
         if (matcher == null) {
             matcher = pattern.matcher(term);
         } else {
             matcher.reset(term);
         }
+        return matcher;
     }
 
-    public static int parseFlags(@Nullable String flagsString) {
+    public List<List<String>> match(String term) {
+        throw new UnsupportedOperationException("method not available");
+    }
+
+    public String replace(String term, String replacement) {
+        throw new UnsupportedOperationException("method not available");
+    }
+
```

Comment:
I'm not sure I like the design of the new abstraction here. 

Is the availability of these two methods depending on the `g` flag? Or why we do need to have these two methods in a base class when they can be unavailable?

Reply:
Well, Pattern/Matcher provide functionality to replace and match, each exposed by a different function (`regexp_replace` and `regexp_matches`). Before both functions shared the same package, but now the later function is a table function, so it has been refactored to a different package.

Both functions accept flags, including `g`, however this flag results in a behaviour not included by `Matcher` and thus needs to be provided by a subclass. I modelled it in this way because I evolved the code from what existed and because the `g` spot is common to both functions.

I will give it another round, to see if I can impress this time.